SPIFBlockDevice.cpp

```diff
@@ -17,368 +17,1019 @@
+    char status_value[2];
+    int status = -1;
+
+    do {
+        if (SPIF_BD_ERROR_OK !=  _spi_send_general_command(SPIF_WREN, SPI_NO_ADDRESS_COMMAND, NULL, 0, NULL, 0)) {
+            tr_error("ERROR:Sending WREN command FAILED\n");
+            break;
+        }
+
+        if ( false == _is_mem_ready()) {
+            tr_error("ERROR: Device not ready, write failed");
+            break;
+        }
+
+        memset(status_value, 0, 2);
+        if (SPIF_BD_ERROR_OK != _spi_send_general_command(SPIF_RDSR, SPI_NO_ADDRESS_COMMAND, NULL, 0, status_value,
+                1)) {   // store received values in status_value
+            tr_error("ERROR: Reading Status Register failed\n");
+            break;
+        }
+
+        if ((status_value[0] & SPIF_STATUS_BIT_WEL) == 0) {
+            tr_error("ERROR: _set_write_enable failed\n");
+            break;
+        }
+        status = 0;
+    } while (false);
+    return status;
 }
 
-bd_size_t SPIFBlockDevice::get_erase_size(bd_addr_t addr) const
+/*********************************************/
+/************* Utility Functions *************/
+/*********************************************/
+int SPIFBlockDevice::_utils_find_addr_region(bd_size_t offset)
 {
-    return SPIF_SE_SIZE;
+    //Find the region to which the given offset belong to
+    if ((offset > _device_size_bytes) || (_regions_count == 0)) {
+        return -1;
+    }
+
+    if (_regions_count == 1) {
+        return 0;
+    }
+
+    for (int i_ind = _regions_count - 2; i_ind >= 0; i_ind--) {
+
+        if (offset > _region_high_boundary[i_ind]) {
+            return (i_ind + 1);
+        }
+    }
+    return -1;
+
 }
 
-bd_size_t SPIFBlockDevice::size() const
+int SPIFBlockDevice::_utils_iterate_next_largest_erase_type(uint8_t& bitfield, int size, int offset, int boundry)
 {
-    if (!_is_initialized) {
-        return BD_ERROR_DEVICE_ERROR;
+    // Iterate on all supported Erase Types of the Region to which the offset belong to.
+    // Iterates from highest type to lowest
+    uint8_t type_mask = ERASE_BITMASK_TYPE4;
+    int i_ind  = 0;
+    int largest_erase_type = 0;
+    for (i_ind = 3; i_ind >= 0; i_ind--) {
+        if (bitfield & type_mask) {
+            largest_erase_type = i_ind;
+            if ( (size > _erase_type_size_arr[largest_erase_type]) &&
+                    ((boundry - offset) > _erase_type_size_arr[largest_erase_type]) ) {
+                break;
+            } else {
+                bitfield &= ~type_mask;
+            }
+        }
+        type_mask = type_mask >> 1;
+    }
+
+    if (i_ind == 4) {
+        tr_error("ERROR: no erase type was found for current region addr");
     }
+    return largest_erase_type;
 
-    return _size;
 }
 
-int SPIFBlockDevice::get_erase_value() const
+/*********************************************/
+/************** Local Functions **************/
+/*********************************************/
+static unsigned int local_math_power(int base, int exp)
```

Comment:
Why not use [std::pow](https://en.cppreference.com/w/cpp/numeric/math/pow)? Then we could potentially share code.

Reply:
I think std::pow operates on float/double and would have required casting (not too critical).. 
Also, using this small local function made it possible for me to avoid including <math.h> or "mbed.h"

Reply:
Oh you're right, I missed that there is no int overload... It was a small thought anyways